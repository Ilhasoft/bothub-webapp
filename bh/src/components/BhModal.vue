<template>
  <transition
    name="fade"
  >
    <div
      v-show="open"
      ref="modal"
      class="bh-modal"
      @click="backgroundClose($event)"
    >
      <bh-card class="bh-modal__content">
        <div class="bh-modal__header">
          <div class="bh-grid">
            <h3 class="bh-modal__header__title bh-grid__item">{{ title }}</h3>
            <button
              ref="closeBtn"
              class="bh-modal__header__close-button clickable"
              @click="close()"
            >
              <bh-icon
                class="bh-modal__close bh-grid__itembh-grid__item--grow-0"
                size="small"
                value="close"
              />
            </button>
          </div>
        </div>
        <slot />
      </bh-card>
    </div>
  </transition>
</template>

<script>
export default {
  name: 'BhModal',
  props: {
    open: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      default: 'Title',
    },
    backgroundCloses: {
      type: Boolean,
      default: true,
    },
  },
  watch: {
    open(value) {
      if (value) {
        this.$bhModal.add(this);
      } else {
        this.$bhModal.rm(this);
      }
    },
  },
  mounted() {
    if (this.open) {
      this.$bhModal.add(this);
    }
  },
  beforeDestroy() {
    this.$bhModal.rm(this);
  },
  methods: {
    close() {
      this.$emit('update:open', false);
    },
    backgroundClose(event) {
      if (this.backgroundCloses && event.target === this.$refs.modal) {
        this.close();
      }
    },
  },
};
</script>

<style lang="scss">
@import '@scss/variables.scss';
@import '@scss/colors.scss';


.bh {
  &-modal {
    position: fixed;
    top: 0;
    left: 0;
    z-index: $modal-z-index;
    width: 100%;
    height: 100%;
    padding: .5rem;
    overflow: auto;
    background-color: rgba(0, 0, 0, .5);

    &__header {
      padding: 1rem;

      &__title {
        text-align: center;
      }

      &__close-button {
        background: none;
        border: none;
        outline: none;
      }
    }

    &__content {
      max-width: 600px;
      margin: 3rem auto;
    }
  }
}
.fade-enter-active,
.fade-leave-active {
  transition: opacity .5s
}

.fade-enter,
.fade-leave-to
{
  opacity: 0
}
</style>
