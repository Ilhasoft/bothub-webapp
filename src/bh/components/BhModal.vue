<template>
  <div
    v-show="open"
    class="bh-modal">
    <div
      class="bh-modal__background"
      @click="backgroundClose()" />
    <div class="bh-modal__close">
      <button
        class="bh-modal__close__button"
        @click="close()">
        <bh-icon
          size="medium"
          value="close" />
      </button>
    </div>
    <bh-card class="bh-modal__content">
      <slot />
    </bh-card>
  </div>
</template>

<script>
export default {
  name: 'BhModal',
  props: {
    open: {
      type: Boolean,
      default: false,
    },
    backgroundCloses: {
      type: Boolean,
      default: true,
    },
  },
  watch: {
    open(value) {
      if (value) {
        this.$bhModal.add(this);
      } else {
        this.$bhModal.rm(this);
      }
    },
  },
  mounted() {
    if (this.open) {
      this.$bhModal.add(this);
    }
  },
  beforeDestroy() {
    this.$bhModal.rm(this);
  },
  methods: {
    close() {
      this.$emit('update:open', false);
    },
    backgroundClose() {
      if (this.backgroundCloses) {
        this.close();
      }
    },
  },
};
</script>

<style lang="scss">
@import '~bh/assets/scss/variables.scss';


.bh {
  &-modal {
    position: fixed;
    top: 0;
    left: 0;
    z-index: $modal-z-index;
    width: 100%;
    height: 100%;

    &__background {
      position: absolute;
      top: 0;
      left: 0;
      z-index: -1;
      width: 100%;
      height: 100%;
      background-color: rgba(0, 0, 0, .5);
    }

    &__close {
      text-align: right;

      &__button {
        padding: .5rem;
        color: white;
        cursor: pointer;
        background: none;
        border: none;
        outline: none;
      }
    }

    &__content {
      max-width: 600px;
      margin: 3rem auto;
    }
  }
}
</style>
