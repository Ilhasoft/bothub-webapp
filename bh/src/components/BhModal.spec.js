/* eslint-env jest */
import { shallowMount, createLocalVue } from '@vue/test-utils';

import BH from '..';
import BhModal from './BhModal';


describe('BhModal.vue', () => {
  let localVue;
  let wrapper;
  beforeEach(() => {
    localVue = createLocalVue();
    localVue.use(BH);
    wrapper = shallowMount(BhModal, {
      localVue,
      propsData: {
        open: true,
      },
    });
  });

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('close modal clicking in closeBtn', () => {
    beforeEach(() => {
      const closeBtn = wrapper.find({ ref: 'closeBtn' });
      closeBtn.trigger('click');
    });

    test('event update:open was emitted', () => {
      expect(wrapper.emitted('update:open')).toBeDefined();
    });

    test('event update:open was emitted with false value', () => {
      const value = wrapper.emitted('update:open')[0][0];
      expect(value).toBeFalsy();
    });

    test('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('close modal clicking in background', () => {
    beforeEach(() => {
      const background = wrapper.find({ ref: 'modal' });
      background.trigger('click');
    });

    test('event update:open was emitted', () => {
      expect(wrapper.emitted('update:open')).toBeDefined();
    });

    test('event update:open was emitted with false value', () => {
      const value = wrapper.emitted('update:open')[0][0];
      expect(value).toBeFalsy();
    });

    test('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('close modal pass false via open prop', () => {
    beforeEach(() => {
      wrapper.setProps({
        open: false,
      });
    });

    test('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe('open modal pass true via open prop', () => {
      beforeEach(() => {
        wrapper.setProps({
          open: true,
        });
      });

      test('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
