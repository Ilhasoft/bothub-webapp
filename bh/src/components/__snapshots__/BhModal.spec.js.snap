// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BhModal.vue close modal clicking in background renders correctly 1`] = `
<div class="bh-modal">
  <div class="bh-modal__close"><button class="bh-modal__close__button">
      <bhicon-stub value="close" size="small"></bhicon-stub>
    </button></div>
  <bhcard-stub class="bh-modal__content"></bhcard-stub>
</div>
`;

exports[`BhModal.vue close modal clicking in closeBtn renders correctly 1`] = `
<div class="bh-modal">
  <div class="bh-modal__close"><button class="bh-modal__close__button">
      <bhicon-stub value="close" size="small"></bhicon-stub>
    </button></div>
  <bhcard-stub class="bh-modal__content"></bhcard-stub>
</div>
`;

exports[`BhModal.vue close modal pass false via open prop open modal pass true via open prop renders correctly 1`] = `
<div class="bh-modal" style="">
  <div class="bh-modal__close"><button class="bh-modal__close__button">
      <bhicon-stub value="close" size="small"></bhicon-stub>
    </button></div>
  <bhcard-stub class="bh-modal__content"></bhcard-stub>
</div>
`;

exports[`BhModal.vue close modal pass false via open prop renders correctly 1`] = `
<div class="bh-modal" style="display: none;">
  <div class="bh-modal__close"><button class="bh-modal__close__button">
      <bhicon-stub value="close" size="small"></bhicon-stub>
    </button></div>
  <bhcard-stub class="bh-modal__content"></bhcard-stub>
</div>
`;

exports[`BhModal.vue renders correctly 1`] = `
<div class="bh-modal">
  <div class="bh-modal__close"><button class="bh-modal__close__button">
      <bhicon-stub value="close" size="small"></bhicon-stub>
    </button></div>
  <bhcard-stub class="bh-modal__content"></bhcard-stub>
</div>
`;
